From 4527d318d473354a9c6ec7a84a3330a131bee596 Mon Sep 17 00:00:00 2001
From: P Dheeraj Srujan Kumar <p.dheeraj.srujan.kumar@intel.com>
Date: Mon, 18 Oct 2021 23:02:00 +0530
Subject: [PATCH] Add Privileges to SseSockets

This commit adds Privileges to Ssesockets.
In the current implementation, once a rule is upgraded (i.e. from
BaseRule to SseSocket), there is no provision to add priviliges.
In this commit, SseSocket inherits PrivilegeParameterTraits to
enable privileges.

Also, in the earlier implementation, .privilege() was called after
BMCWEB_ROUTE(). This results in adding those privileges to the Base
rule that is created. By moving the privileges() below websocket(),
the privileges are applied to the Ssesocket.

Tested:
 - SSE Subscription was successful with Admin and Operator Users
 - SSE Subscription was rejected while using Readonly User
 - websocket_test.py Passed
 - Admin and Operator users were able to access KVM on WebUI
 - Readonly User was unable to access KVM on WebUI

Change-Id: I41739401893b1c2bf718e11ec7676d69f954c98f
Signed-off-by: P Dheeraj Srujan Kumar <p.dheeraj.srujan.kumar@intel.com>
Signed-off-by: A Sri Anjaneyulu <srix.anjaneyulu.alapati@intel.com>
---
 http/routing.hpp             | 4 +++-
 include/eventservice_sse.hpp | 2 ++
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/http/routing.hpp b/http/routing.hpp
index 43621045..fc938d5c 100644
--- a/http/routing.hpp
+++ b/http/routing.hpp
@@ -408,7 +408,9 @@ class WebSocketRule :
     std::function<void(crow::websocket::Connection&)> errorHandler;
 };
 
-class SseSocketRule : public BaseRule
+class SseSocketRule :
+    public BaseRule,
+    public PrivilegeParameterTraits<SseSocketRule>
 {
     using self_t = SseSocketRule;
 
diff --git a/include/eventservice_sse.hpp b/include/eventservice_sse.hpp
index 7d17b303..c872eab8 100644
--- a/include/eventservice_sse.hpp
+++ b/include/eventservice_sse.hpp
@@ -192,6 +192,8 @@ inline void requestRoutes(App& app)
 {
     BMCWEB_ROUTE(app, "/redfish/v1/EventService/SSE/")
         .serverSentEvent()
+        .privileges(redfish::privileges::
+                        privilegeSetConfigureManagerOrConfigureComponents)
         .onopen([](std::shared_ptr<crow::SseConnection>& conn,
                    const crow::Request& req, crow::Response& res) {
             BMCWEB_LOG_DEBUG << "Connection " << conn << " opened.";
-- 
2.17.1

